Use `i18next` for UI strings and add new `injectTranslations` plugin callback

HiDeoo opened the pull request:
**This PR is currently based on #1841.**

#### Description

- Closes #1496 and also relates to the #962 proposal

This PR:

- Adds a new `injectTranslations` plugin callback for plugins to inject translations to Starlight.
- Refactors the UI strings/translations to a middleware.
- Refactors the UI strings/translations to use [i18next](https://www.i18next.com/translation-function/plurals).
- Exposes UI strings/translations API to users and plugins.

Note that the links validation is failing a lot but this is expected as the links validator plugin does not support the configuration syntax based on #1841.

#### Remaining tasks

- [x] Figure out typings for plugins accessing their own custom injected translations (error in `examples/custom-plugin/MarkdownContent.astro`)
- [x] Improve the documentation which is currently a very rough draft just to document the available APIs
- [x] Bump peer dependencies
	- [x] `starlight`: Astro `4.14.0` for `injectTypes()`
	- [x] `starlight-docsearch`: Latest Starlight as it uses this new feature
- [x] Add a changeset
  - [x] This would span across multiple packages, e.g. `@astrojs/starlight-docsearch`
- [x] See if some tests can be removed, e.g. the ones that used to check for `Astro.props.labels` which now checks that it's not defined anymore (useful during dev but maybe not worth keeping)
- [x] Address and remove all the `// TODO(HiDeoo)` comments
- [x] Remove debug content
  - [x] `docs/src/content/docs/guides/debug-i18next.mdx`
  - [x] `docs/src/components/i18next-debug.astro`
  - [x] `docs/src/content/i18n/**/*`
  - [x] `examples/custom-plugin/**/*`

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files


| Locale | File                                                                                                                                                                      | Note                                                       |
| ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| en     | [guides/debug-i18next.mdx](https://github.com/withastro/starlight/blob/f576a19f28440ff943500c9ae94eba6b24b50e38/docs%2Fsrc%2Fcontent%2Fdocs%2Fguides%2Fdebug-i18next.mdx) | Source added, will be tracked.                             |
| en     | [guides/i18n.mdx](https://github.com/withastro/starlight/blob/f576a19f28440ff943500c9ae94eba6b24b50e38/docs%2Fsrc%2Fcontent%2Fdocs%2Fguides%2Fi18n.mdx)                   | Source changed, localizations will be marked as outdated.  |
| en     | [reference/overrides.md](https://github.com/withastro/starlight/blob/f576a19f28440ff943500c9ae94eba6b24b50e38/docs%2Fsrc%2Fcontent%2Fdocs%2Freference%2Foverrides.md)     | Source changed, localizations will be marked as outdated.  |
| en     | [reference/plugins.md](https://github.com/withastro/starlight/blob/f576a19f28440ff943500c9ae94eba6b24b50e38/docs%2Fsrc%2Fcontent%2Fdocs%2Freference%2Fplugins.md)         | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

delucis commented:
Back for my weekly round of update branch and break things 😁 

This one’s actually a good one: the new sidebar feature added in #1874 revealed an inconsistency when switching to Astro’s locale. It works on all pages **except** for our 404 page.

- On regular English pages: `locale` is `undefined` as expected
- On the 404 page: `locale` is `'en'`

This means constructing the slug in the sidebar logic breaks because it’s looking for `en/getting-started` instead of `getting-started`.

Haven’t dug into where that logic happens yet, but thought I’d quickly explain myself for breaking things!

delucis commented on packages/starlight/utils/route-data.ts:
```diff
@@ -34,8 +34,6 @@ export interface StarlightRouteData extends Route {
 	editUrl: URL | undefined;
-	/** Record of UI strings localized for the current page. */
-	labels: ReturnType<ReturnType<typeof useTranslations>['all']>;
 }
```
Making a quick note before I forget of an idea to maybe make migration a bit easier for people.

1. Keep the type here and mark it as deprecated.

2. Create a proxy that throws a helpful error when accessing `labels`.

**N.B. `AstroError` will need adding to imports**

```suggestion
	/** @deprecated Use `Astro.locals.t()` instead **/
	labels: Record<string, never>;
}

// TODO: remove in a future release once people have updated. No later than v1.
const labels: Record<string, never> = new Proxy(
	{},
	{
		get(_, key) {
			const label = String(key);
			throw new AstroError(
				`The \`labels\` prop in component overrides has been removed.`,
				`Replace \`labels["${label}"]\` with \`Astro.locals.t("${label}")\` instead.\n` +
					'For more information see https://starlight.astro.build/guides/i18n/#accessing-ui-strings'
			);
		},
	}
);
```

delucis commented on packages/starlight/utils/route-data.ts:
```diff
@@ -58,7 +56,6 @@ export function generateRouteData({
 		lastUpdated: getLastUpdated(props),
 		editUrl: getEditUrl(props),
-		labels: useTranslations(locale).all(),
 	};
```
Then we can include that proxy in the route data:

```suggestion
		labels,
	};
```

HiDeoo commented:
Updated the branch and also added support for [`injectTypes()`](https://docs.astro.build/en/reference/integrations-reference/#injecttypes-options) to automatically inject plugin UI string type definitions.

delucis commented on packages/starlight/__tests__/i18n-non-root-single-locale/translations-fs.test.ts:
```diff
@@ -9,7 +9,8 @@ describe('createTranslationSystemFromFs', () => {
 			// Using non-existent `_src/` to ignore custom files in this test fixture.
-			{ srcDir: new URL('./_src/', import.meta.url) }
+			{ srcDir: new URL('./_src/', import.meta.url) },
+			{}
```
Should we make this third parameter optional with a default of `{}` like `createTranslationSystem()` itself to avoid all these pointless extra `{}` in tests? Or do you think it’s too risky we’d forget to pass the parameter in internal usage?

delucis commented on packages/starlight/utils/createTranslationSystem.ts:
```diff
@@ -1,27 +1,60 @@
+	const i18n = i18next.createInstance();
+	i18n.init({
+		resources: translations,
+		fallbackLng:
```
Why did they think omitting **ONE** letter was a good idea? 😂 
Screams “typo” at my brain 😅

delucis reviewed (commented):
Looking great after the latest `injectTypes()` changes! Very excited for this.

Shared some quick notes. I haven’t tackled the docs part yet, but I think it’s actually looking in a good place already.

HiDeoo commented:
Pushed a fix for the type issues of a plugin accessing their own injected UI strings. My next step will be documenting all this from a plugin author perspective in the [`injectTranslations()`](<https://deploy-preview-1923--astro-starlight.netlify.app/reference/plugins/#injecttranslations>) docs and also take care of your review comments.

HiDeoo commented on packages/starlight/locals.d.ts:
```diff
@@ -0,0 +1,19 @@
+	export interface StarlightLocals {
+		t: import('./utils/createTranslationSystem').I18nT;
+	}
+}
```
The triple-slash directives refactor includes some comments which hopefully makes the whole architecture more understandable. Let me know if it's not the case.

HiDeoo commented on packages/starlight/utils/translations-fs.ts:
```diff
@@ -11,12 +11,13 @@ import type { AstroConfig } from 'astro';
 ) {
 	/** All translation data from the i18n collection, keyed by `id`, which matches locale. */
-	let userTranslations: Record<string, i18nSchemaOutput> = {};
+	let userTranslations: Record<string, i18nSchemaOutput> = { ...pluginTranslations };
```
Definitely not, I would guess early exploration I forgot to refactor now that `createTranslationSystem` accepts the plugin translations directly. This should now be fixed.

HiDeoo commented:
Updated the branch with changes that should address the review comments and everything we talked during the Talking & Doc'ing special. I also added new developer-focused documentation for `injectTranslations()` where I tried to mimic the ["Middleware types"](https://docs.astro.build/en/guides/middleware/#middleware-types) guide but I'm not sure if it's any good, or if we should go into more detail or something.

delucis reviewed (commented):
Quick note on those `.d.ts` files. Rest of the code LGTM! Will take a look at the docs now.

HiDeoo commented on packages/starlight/index.ts:
```diff
@@ -1,3 +1,18 @@
+ * These triple-slash directives defines dependencies to various declaration files that will be
+ * loaded when a user imports the Starlight integration in their Astro configuration file. These
+ * directives must be first at the top of the file and can only be preceded by this comment.
+ *
+ * - `./locals.d.ts` registers the various Astro middlewares used by Starlight and also defines a
+ * 	 namespace reserved for Starlight (only used for i18n at the moment) that can be extended by
+ *   plugins using module augmentation and interface merging.
+ * - `./i18n.d.ts` imports the original `i18next` types and extends to configure the Starlight
+ *   namespace (note that this can only be done in a different file due to a top-level `import`
+ *   making it non-ambient).
```
My initial thought was to have everything in this file to avoid showing these comments to a user when they "Go to Definition", e.g. for `locals.t`, but now seeing the comments with your changes, I think it's fine and I made it more of an issue than it actually is. I'll update the PR tomorrow morning :thumbsup:

delucis replied:
Ah yeah didn’t consider these files as user-facing. But still, I agree, I think these comments are OK also there (plus not sure quite how often people will click through). If it turns out to be more frequent, we could even consider something more tailored to that audience.

HiDeoo commented:
> Also, small side note: I noticed that the `// @ts-expect-error` in the docs Astro config is only needed when type checking from the root. Running `pnpm astro check` from the `docs/` directory directly doesn’t trigger any errors interestingly.

Which `// @ts-expect-error` are you referring to? I cannot find one in the `docs/astro.config.mjs` file 🤔

delucis commented:
> Which `// @ts-expect-error` are you referring to? I cannot find one in the `docs/astro.config.mjs` file 🤔

🤦 — wrong PR! That’s on the component documentation PR but I got myself confused 😅

HiDeoo commented:
> > Which `// @ts-expect-error` are you referring to? I cannot find one in the `docs/astro.config.mjs` file 🤔
> 
> 🤦 — wrong PR! That’s on the component documentation PR but I got myself confused 😅

Oh, weird, I always use `pnpm typecheck` at the root so didn't realize but I'll take a look. I hope I won't need to use `ts-ignore` (casting in JSDoc as the type is not exported from anywhere is even more annoying).

HiDeoo commented:
Managed to understand the grammar type incompatibility and fixed it in the other PR at the grammar generator level so we no longer need the `@ts-expect-error` annotation in the docs Astro config.

delucis reviewed (commented):
Note for the future: I think it would be nice to support keys in user i18n files that are not explicitly set in the i18n collection schema. This might initially seem like an odd idea, but particularly for plugins I think it will make for a nicer DX. Instead of having to merge in a custom schema the plugin exports, you could just set the keys in your files and that’s it. Types are already provided by plugins, so there’s no real advantage to providing the types _again_ in the schema.

To do that we’d want to add [`.passthrough()`](https://zod.dev/?id=passthrough) to the default i18n schema. That seems simple enough, but because of type nonsense I couldn’t quickly get it to line up with our types without using an `as unknown` which feels wrong. (Does work though! Just the types are wonky 😅)

To go one step further: there’s even an argument for dropping content collections entirely here and switching to our own type generation now that that is easier for Astro integrations. That way a user could also define custom keys in their i18n files and get types in `Astro.locals.t()` without needing to also update the schema, i.e. a start-up, use `node:fs` to load i18n files, generate types from the keys present, and then expose the dictionaries as a virtual module.

Anyway, very much **for the future**, but wanted to share here so it’s not just in my brain.

delucis reviewed (commented):
Left some notes on the reference page. I think we can also start cleaning up the TODOs etc. Would you mind if I dive in on the i18n guide directly? (I can also help with the TODOs if you like, but would wait for you to ask that as undoubtedly you have a better handle on them than me.)

HiDeoo commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -161,3 +162,50 @@ The example above will log a message that includes the provided info message:
+To type the `locals.t` object in the context of your plugin, declare the following global namespaces in a TypeScript declaration file:
+
+```ts
+// env.d.ts
````
By default, no, just like our `virtual.d.ts` file for example where we don't do anything special for it. You could definitely setup your TS config so that it does not, but in this case, I assume it would be up to the author to handle that? I'm also not confident we could even document all cases.

delucis replied:
OK, perfect. If nothing else in theory is needed, no need to add anything. That’s great.

HiDeoo commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -161,3 +162,50 @@ The example above will log a message that includes the provided info message:
+    'myPlugin.doThing': string;
+  }
+}
+```
````
Yeah, I was a bit hesitant too to describe more advanced setup (which are easier in the end) but looking at your great intro sentence and the small 2 examples, I think it's worth it 👍

HiDeoo replied:
And yes, I think the simplified syntax should work too and I made it a bit more complex than necessary in the example. I'll still test it and if it works, I'll update your suggestion.

HiDeoo commented on .changeset/cool-experts-sort.md:
```diff
@@ -0,0 +1,19 @@
+
+Overhauls the built-in localization system which is now powered by the [`i18next`](https://www.i18next.com/) library and available to use anywhere in your documentation website.
+
+See the [“Accessing UI strings”](https://starlight.astro.build/guides/i18n/#using-ui-translations) guide to learn more about how to access built-in UI labels or your own custom strings in your project. Plugin authors can also use the new [`injectTranslations()`](https://starlight.astro.build/reference/plugins/#injecttranslations) helper to add or update translation strings.
```
```suggestion
See the [“Using UI translations”](https://starlight.astro.build/guides/i18n/#using-ui-translations) guide to learn more about how to access built-in UI labels or your own custom strings in your project. Plugin authors can also use the new [`injectTranslations()`](https://starlight.astro.build/reference/plugins/#injecttranslations) helper to add or update translation strings.
```

Should we also update the title with the latest changes?

delucis replied:
Oh definitely. Find and replace comes back to bite me 😁

HiDeoo commented on docs/src/content/docs/reference/plugins.md:
````diff
@@ -161,3 +162,71 @@ The example above will log a message that includes the provided info message:
+};
+```
+
+To use the injected translations in your plugin UI, follow the [“Accessing UI strings” guide](/guides/i18n/#using-ui-translations).
````
```suggestion
To use the injected translations in your plugin UI, follow the [“Using UI translations” guide](/guides/i18n/#using-ui-translations).
```

And here too maybe?

sarah11918 reviewed (approved):
Y'all are just too good at this! I am so happy to see this accomplishment, and really proud to be associated with the quality of work in this project!

One tiny thought I had while reading the changeset, but the docs are fantastic.

delucis reviewed (approved):
Everything looks good to me! This is a very exciting feature to be added. Excited to see what people do with and excited to see if we got anything wrong 😁 

Thank you again @HiDeoo — all yours to merge! 🌟
